R/add_labels.R

```diff
@@ -0,0 +1,157 @@
+#' @title Add variable and value labels
+#' @name add_labels
+#'
+#' @description
+#' Add variable and values labels to a variable or variables in a data frame.
+#' Labels are stored as attributes (`"label"` for variable labels and `"labels"`)
+#' for value labels.
+#'
+#' @param x A data frame, factor or vector.
+#' @param variable The variable label as string.
+#' @param values The value labels as (named) character vector. If `values` is
+#' *not* a named vector, the length of lables must be equal to the length of
+#' unique values. For a named vector, the left-hand side (LHS) is the value in
+#' `x`, the right-hand side (RHS) the associated value label. Non-matching
+#' labels are omitted.
+#' @param ... Currently not used.
+#' @inheritParams find_columns
+#'
+#' @inheritSection center Selection of variables - the `select` argument
+#'
+#' @return A labelled variable, or a data frame of labelled variables.
+#'
+#' @examples
+#' x <- 1:3
+#' # labelling by providing required number of labels
+#' add_labels(
+#'   x,
+#'   variable = "My x",
+#'   values = c("one", "two", "three")
+#' )
+#'
+#' # labelling using named vectors
+#' data(iris)
+#' out <- add_labels(
+#'   iris$Species,
+#'   variable = "Labelled Species",
+#'   values = c(`setosa` = "Spec1", `versicolor` = "Spec2", `virginica` = "Spec3")
+#' )
+#' str(out)
+#'
+#' # Partial labelling
+#' x <- 1:5
+#' add_labels(
+#'   x,
+#'   variable = "My x",
+#'   values = c(`1` = "lowest", `5` = "highest"),
+#'   verbose = FALSE
+#' )
+#' @export
+add_labels <- function(x, ...) {
+  UseMethod("add_labels")
+}
+
+
+#' @export
+add_labels.default <- function(x, verbose = TRUE, ...) {
+  if (isTRUE(verbose)) {
+    insight::format_alert(
+      sprintf("Adding labels currently not possible for variables of class `%s`.", class(x)[1])
+    )
+  }
+  x
+}
+
+#' @export
+add_labels.numeric <- function(x, variable = NULL, values = NULL, verbose = TRUE, ...) {
+  # add variable label
+  if (!is.null(variable)) {
+    if (is.character(variable) && length(variable) == 1) {
+      attr(x, "label") <- variable
+    } else if (verbose) {
+      insight::format_alert("Variable labels must be a character string.")
+    }
+  }
+
+  # extract unique values
+  unique_values <- as.vector(sort(stats::na.omit(unique(x))))
+  labels <- NULL
+
+  # do we have a names vector for "values"?
+  # else check if number of labels and values match
+  if (is.null(names(values))) {
+    if (length(values) == length(unique_values)) {
+      labels <- stats::setNames(unique_values, values)
+    } else if (verbose) {
+      insight::format_alert(
+        "Cannot add labels. Number of unique values and number of value labels are not equal.",
+        sprintf("There are %i unique values and %i provided labels.", length(unique_values), length(values))
+      )
+    }
+  } else {
+    # check whether we have matches of labels and values
+    matching_labels <- names(values) %in% unique_values
```

Comment:
perhaps throw an error if all values are not named, it would be clearer than this:
```r
> add_labels(c(1, 2), values = c("foo", `1` = "lowest"))
Following labels were associated with values that don't exist:
  `foo ()`
[1] 1 2
attr(,"labels")
lowest 
     1 
```